sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/ContextualServer.java

```diff
@@ -0,0 +1,109 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugin.typescript;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import org.sonar.api.Startable;
+import org.sonar.api.batch.InstantiationStrategy;
+import org.sonar.api.config.Configuration;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.plugin.typescript.executable.ExecutableBundle;
+import org.sonar.plugin.typescript.executable.ExecutableBundleFactory;
+import org.sonarsource.api.sonarlint.SonarLintSide;
+
+@InstantiationStrategy(InstantiationStrategy.PER_BATCH)
+@SonarLintSide
+public class ContextualServer implements Startable {
+
+  // SonarLint should pass in this property an absolute path to the directory containing TypeScript dependency
+  private static final String TYPESCRIPT_DEPENDENCY_LOCATION_PROPERTY = "sonar.typescript.internal.typescriptLocation";
+
+  private static final Logger LOG = Loggers.get(ContextualServer.class);
+
+  private Configuration configuration;
+  private ExecutableBundleFactory bundleFactory;
+  private static AtomicReference<Process> serverProcess = new AtomicReference<>();
+
+  public ContextualServer(ExecutableBundleFactory bundleFactory) {
+    this(null, bundleFactory);
+  }
+
+  public ContextualServer(Configuration configuration, ExecutableBundleFactory bundleFactory) {
+    this.configuration = configuration;
+    this.bundleFactory = bundleFactory;
+  }
+
+  @Override
+  public void start() {
+    if (serverProcess.get() != null && serverProcess.get().isAlive()) {
+      LOG.debug("Skipping SonarTS Server start, already running");
+      return;
+    }
+
+    LOG.warn("Attempting SonarTS Server start");
+
+    if (configuration == null) {
+      LOG.warn("Skipping SonarTS Server start due to null configuration");
+      return;
+    }
+
+    final ExecutableBundle bundle = bundleFactory.createAndDeploy(getServerDir(), configuration);
+    ProcessBuilder processBuilder = new ProcessBuilder(bundle.getSonarTSServerCommand().commandLineTokens());
+
+    Optional<String> typescriptLocation = configuration.get(TYPESCRIPT_DEPENDENCY_LOCATION_PROPERTY);
+    if (typescriptLocation.isPresent()) {
+      SensorContextUtils.setNodePath(new File(typescriptLocation.get()), processBuilder);
+
+    } else {
+      LOG.error("No value provided by SonarLint for TypeScript location; property " + TYPESCRIPT_DEPENDENCY_LOCATION_PROPERTY);
+    }
+
+    try {
+      processBuilder.inheritIO();
+      serverProcess.set(processBuilder.start());
+      synchronized (this) { //TODO remove this and find
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=sonarlint-integration-poc&issues=AWDgV_ClbBlXo5JPxi1c&open=AWDgV_ClbBlXo5JPxi1c)

[ISSUE_KEY]: AWDgV_ClbBlXo5JPxi1c